File: algorithmic_toolbox/week2_algorithmic_warmup/5_fibonacci_number_again/fibonacci_huge.cpp
#include <iostream>

long long get_fibonacci_huge_naive(long long n, long long m) {
  if (n <= 1)
    return n;

  long long previous = 0;
  long long current = 1;

  for (long long i = 0; i < n - 1; ++i) {
    long long tmp_previous = previous;
    previous = current;
    current = (tmp_previous + current) % m;
  }

  return current;
}

long long get_pisano_period(long long m) {
  long long a = 0, b = 1, c = a + b;

  for (int i = 0; i < m * m; i++) {
    c = (a + b) % m;
    a = b;
    b = c;

    if (a == 0 && b == 1) return i + 1;
  }

  throw std::runtime_error("could not find pisano period");
}

long long get_fibonacci_huge(long long n, long long m) {
  return get_fibonacci_huge_naive(n % get_pisano_period(m), m);
}

int main() {
  long long n, m;
  std::cin >> n >> m;

  std::cout << get_fibonacci_huge(n, m) << std::endl;
  std::cout << get_fibonacci_huge_naive(n, m) << std::endl;

  return 0;
}

File: algorithmic_toolbox/week2_algorithmic_warmup/5_fibonacci_number_again/FibonacciHuge.java
import java.util.Scanner;

public class FibonacciHuge {
  private static long getFibonacciHugeNaive(long n, long m) {
    if (n <= 1)
      return n;

    long previous = 0;
    long current = 1;

    for (long i = 0; i < n - 1; ++i) {
      long tmp_previous = previous;
      previous = current;
      current = (tmp_previous + current) % m;
    }

    return current;
  }

  private static long getPisanoPeriod(long m) throws Exception {
    long a = 0, b = 1, c = a + b;

    for (int i = 0; i < m * m; i++) {
      c = (a + b) % m;
      a = b;
      b = c;

      if (a == 0 && b == 1)
        return i + 1;
    }

    throw new Exception("could not find pisano period");
  }

  private static long getFibonacciHuge(long n, long m) throws Exception {
    return getFibonacciHugeNaive(n % getPisanoPeriod(m), m);
  }

  public static void main(String[] args) throws Exception {
    Scanner scanner = new Scanner(System.in);
    long n = scanner.nextLong();
    long m = scanner.nextLong();
    scanner.close();

    System.out.println(getFibonacciHuge(n, m));
    System.out.println(getFibonacciHugeNaive(n, m));
  }
}
